odl/discr/tensor_ops.py

```diff
@@ -112,6 +117,479 @@ def base_space(self):
+         [0.0, 1.0, 2.0, 3.0, 4.0]]
+        """
+        if out is None:
+            out = self.range.element()
+
+        # TODO: this pipes CUDA arrays through NumPy. Write native operator.
+        out_arr = out.asarray()
+        finite_diff(x.asarray(), out=out_arr, axis=self.axis, dx=self.dx,
+                    method=self.method, padding_method=self.padding_method,
+                    padding_value=self.padding_value,
+                    edge_order=self.edge_order)
+
+        # self assignment: no overhead in the case out_arr is a view
+        out[:] = out_arr
+        return out
+
+    @property
+    def adjoint(self):
+        """Return the adjoint operator."""
+        return -self
+
+
+class Gradient(PointwiseTensorFieldOperator):
+
+    """Spatial gradient operator for `DiscreteLp` spaces.
+
+    Calls helper function `finite_diff` to calculate each component of the
+    resulting product space vector. For the adjoint of the `Gradient`
+    operator, zero padding is assumed to match the negative `Divergence`
+    operator
+    """
+
+    def __init__(self, domain=None, range=None, method='forward'):
+        """Initialize a `Gradient` operator instance.
+
+        Zero padding is assumed for the adjoint of the `Gradient`
+        operator to match  negative `Divergence` operator.
+
+        Parameters
+        ----------
+        domain : `DiscreteLp`, optional
+            The space of elements which the operator acts on.
+            This is required if ``range`` is not given.
+        range : power space of `DiscreteLp`, optional
+            The space of elements to which the operator maps.
+            This is required if ``domain`` is not given.
+        method : {'central', 'forward', 'backward'}, optional
+            Finite difference method to be used
+
+        Examples
+        --------
+        >>> import odl
+        >>> dom = odl.uniform_discr([0, 0], [1, 1], (10, 20))
+        >>> ran = odl.ProductSpace(dom, dom.ndim)  # 2-dimensional
+        >>> grad_op = Gradient(dom)
+        >>> grad_op.range == ran
+        True
+        >>> grad_op2 = Gradient(range=ran)
+        >>> grad_op2.domain == dom
+        True
+        >>> grad_op3 = Gradient(domain=dom, range=ran)
+        >>> grad_op3.domain == dom
+        True
+        >>> grad_op3.range == ran
+        True
+        """
+        if domain is None and range is None:
+            raise ValueError('either domain or range must be specified.')
+
+        if domain is None:
+            if not isinstance(range, ProductSpace):
+                raise TypeError('range {!r} is not a ProductSpace instance.'
+                                ''.format(range))
+            domain = range[0]
+
+        if range is None:
+            if not isinstance(domain, DiscreteLp):
+                raise TypeError('domain {!r} is not a `DiscreteLp` '
+                                'instance.'.format(domain))
+            range = ProductSpace(domain, domain.ndim)
+
```

Comment:
The actual checks happen in the parent class, except for the `DiscreteLp` thing. Otherwise I only test to make sure I can use the attributes I need.